File: CS/TreeListCellMerging/MyTreeListViewInfo.cs
// Developer Express Code Central Example:
// How to add a horizontal cells merge capability to an XtraTreeList control
// 
// This sample shows how to add a horizontal cell merge capability to an
// XtraTreeList control.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E2501

using System.Collections;
using System.Drawing;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Columns;
using DevExpress.XtraTreeList.Nodes;
using DevExpress.XtraTreeList.ViewInfo;

namespace TreeListCellMerging
{
	public class MyTreeListViewInfo : TreeListViewInfo
	{
		public MyTreeListViewInfo(TreeList fTreeList)
			: base(fTreeList)
		{
		}

        

		protected string GetCellDisplayText(RowInfo ri, int cellIndex)
		{
			TreeListColumn col = ((CellInfo)ri.Cells[cellIndex]).Column;
			return ri.Node.GetDisplayText(col);
		}

		protected string GetPrevNodeCellDisplayText(RowInfo ri, int cellIndex)
		{
			TreeListNode prevNode = ri.Node.PrevNode;
			if ( prevNode == null )
				return "";

			TreeListColumn col = ((CellInfo)ri.Cells[cellIndex]).Column;
			return prevNode.GetDisplayText(col);
		}
        

		protected override void CalcRowCellsInfo(RowInfo ri, ArrayList viewInfoList)
		{
			base.CalcRowCellsInfo(ri, viewInfoList);
			if ( !((MyTreeList)TreeList).OptionsView.AllowHorizontalMerge )
				return;

			if ( TreeList.OptionsSelection.EnableAppearanceFocusedRow )
				TreeList.OptionsSelection.EnableAppearanceFocusedRow = false;

			for ( int i = ri.Cells.Count - 1; i > 0; i-- )
			{
				string prevDisplayText = GetCellDisplayText(ri, i - 1);
				string currDisplayText = GetCellDisplayText(ri, i);

				if ( prevDisplayText == currDisplayText )
				{
					CellInfo prevCell = (CellInfo)ri.Cells[i - 1];
					CellInfo currCell = (CellInfo)ri.Cells[i];

					AllowMergeRowCellsEventArgs args = new AllowMergeRowCellsEventArgs(ri, i, i - 1);
					((MyTreeList)TreeList).RaiseMergeRowCells(args);
					if ( !args.Merge )
						continue;

					for ( int j = ri.Lines.Count - 1; j >= 0; j-- )
					{
						LineInfo lineInfo = (LineInfo)ri.Lines[j];
						if ( lineInfo.Appearance != PaintAppearance.VertLine )
							continue;

						if ( lineInfo.Rect.X == prevCell.Bounds.X + prevCell.Bounds.Width )
						{
							ri.Lines.RemoveAt(j);
							break;
						}
					}

                    Rectangle mergedBounds = new Rectangle(prevCell.Bounds.X, prevCell.Bounds.Y, currCell.Bounds.Right - prevCell.Bounds.X, prevCell.Bounds.Height);
                    mergedBounds.Inflate(-1, -1);
                    prevCell.CalcViewInfo(GInfo.Graphics, Point.Empty, mergedBounds);
                    prevCell.EditorViewInfo.Bounds = mergedBounds;

					ri.Cells.RemoveRange(i - 1, 2);
					ri.Cells.Insert(i - 1, prevCell);
				}
			}
		}
	}
}

File: VB/TreeListCellMerging/MyTreeListViewInfo.vb
' Developer Express Code Central Example:
' How to add a horizontal cells merge capability to an XtraTreeList control
' 
' This sample shows how to add a horizontal cell merge capability to an
' XtraTreeList control.
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E2501


Imports Microsoft.VisualBasic
Imports System.Collections
Imports System.Drawing
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Columns
Imports DevExpress.XtraTreeList.Nodes
Imports DevExpress.XtraTreeList.ViewInfo

Namespace TreeListCellMerging
	Public Class MyTreeListViewInfo
		Inherits TreeListViewInfo
		Public Sub New(ByVal fTreeList As TreeList)
			MyBase.New(fTreeList)
		End Sub



		Protected Function GetCellDisplayText(ByVal ri As RowInfo, ByVal cellIndex As Integer) As String
			Dim col As TreeListColumn = (CType(ri.Cells(cellIndex), CellInfo)).Column
			Return ri.Node.GetDisplayText(col)
		End Function

		Protected Function GetPrevNodeCellDisplayText(ByVal ri As RowInfo, ByVal cellIndex As Integer) As String
			Dim prevNode As TreeListNode = ri.Node.PrevNode
			If prevNode Is Nothing Then
				Return ""
			End If

			Dim col As TreeListColumn = (CType(ri.Cells(cellIndex), CellInfo)).Column
			Return prevNode.GetDisplayText(col)
		End Function


		Protected Overrides Sub CalcRowCellsInfo(ByVal ri As RowInfo, ByVal viewInfoList As ArrayList)
			MyBase.CalcRowCellsInfo(ri, viewInfoList)
			If Not(CType(TreeList, MyTreeList)).OptionsView.AllowHorizontalMerge Then
				Return
			End If

			If TreeList.OptionsSelection.EnableAppearanceFocusedRow Then
				TreeList.OptionsSelection.EnableAppearanceFocusedRow = False
			End If

			For i As Integer = ri.Cells.Count - 1 To 1 Step -1
				Dim prevDisplayText As String = GetCellDisplayText(ri, i - 1)
				Dim currDisplayText As String = GetCellDisplayText(ri, i)

				If prevDisplayText = currDisplayText Then
					Dim prevCell As CellInfo = CType(ri.Cells(i - 1), CellInfo)
					Dim currCell As CellInfo = CType(ri.Cells(i), CellInfo)

					Dim args As New AllowMergeRowCellsEventArgs(ri, i, i - 1)
					CType(TreeList, MyTreeList).RaiseMergeRowCells(args)
					If (Not args.Merge) Then
						Continue For
					End If

					For j As Integer = ri.Lines.Count - 1 To 0 Step -1
						Dim lineInfo As LineInfo = CType(ri.Lines(j), LineInfo)
						If lineInfo.Appearance IsNot PaintAppearance.VertLine Then
							Continue For
						End If

						If lineInfo.Rect.X = prevCell.Bounds.X + prevCell.Bounds.Width Then
							ri.Lines.RemoveAt(j)
							Exit For
						End If
					Next j

					Dim mergedBounds As New Rectangle(prevCell.Bounds.X, prevCell.Bounds.Y, currCell.Bounds.Right - prevCell.Bounds.X, prevCell.Bounds.Height)
					mergedBounds.Inflate(-1, -1)
					prevCell.CalcViewInfo(GInfo.Graphics, Point.Empty, mergedBounds)
					prevCell.EditorViewInfo.Bounds = mergedBounds

					ri.Cells.RemoveRange(i - 1, 2)
					ri.Cells.Insert(i - 1, prevCell)
				End If
			Next i
		End Sub
	End Class
End Namespace
